### PR TITLE
Add possibility to remove option values from Product Option

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A
|-----------------|-----
| Branch?         | 2.0 
| New feature?    | a bit
| BC breaks?      | yes
| License         | MIT

Screenshot 2024-09-27 at 17 41 46

### CODE REVIEW

File: features/admin/product/managing_product_options/managing_values_of_product_option.feature [hunk 1]
```diff
@@ -37,3 +37,16 @@ Feature: Managing option values of a product option
         Then I should be notified that it has been successfully edited
         And this product option should not have the "M" option value
         And this product option should have the "L" option value
+
+    @api @ui @mink:chromedriver
+    Scenario: Removing product option value that is in use by product variant
+        And the store has a "Car" configurable product
```
Comment (GSadee) [inline features/admin/product/managing_product_options/managing_values_of_product_option.feature]:
```suggestion
        Given the store has a "Car" configurable product
```

File: features/admin/product/managing_product_options/managing_values_of_product_option.feature [hunk 2]
```diff
@@ -37,3 +37,16 @@ Feature: Managing option values of a product option
...
+        And this product has option "Model" with values "Sedan", "Kombi" and "Cabrio"
+        And this product has "Car-Variant-1" variant priced at "$20.00" configured with "Sedan" option value
+        And this product has "Car-Variant-2" variant priced at "$25.00" configured with "Kombi" option value
+        And this product has "Car-Variant-3" variant priced at "$50.00" configured with "Cabrio" option value
+        When I want to modify the "Model" product option
+        And I delete the "Sedan" option value of this product option
+        And I try to save my changes
+        Then I should be notified that it is in use
+        And product option "Model" should have the "Sedan" option value
```
Comment (GSadee) [inline features/admin/product/managing_product_options/managing_values_of_product_option.feature]:
```suggestion
        And product option "Model" should still have the "Sedan" option value
```

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductOptionsContext.php [hunk 3]
```diff
@@ -139,11 +139,11 @@ public function iApplyToAllTheOptionValueIdentifiedBy(string $code, string $loca
     }
 
     /**
-     * @When I delete the :value option value of this product option
+     * @When I delete the :optionValue option value of this product option
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductOptionsContext.php]: Probably this change should be reverted

File: src/Sylius/Behat/Element/Admin/ProductOption/FormElement.php [hunk 4]
```diff
@@ -60,6 +71,7 @@ protected function getDefinedElements(): array
             'last_option_value' => '[data-test-option-values] [data-test-option-value]:last-child',
             'name' => '#sylius_admin_product_option_translations_%locale_code%_name',
             'option_value' => '[data-test-option-values] input[id$="_translations_%locale_code%_value"][value="%option_value%"]',
+            'delete_option_value' => '[data-test-delete-option-value]',
```
Comment (GSadee) [inline src/Sylius/Behat/Element/Admin/ProductOption/FormElement.php]: Add this key in alphabetical order

File: src/Sylius/Bundle/AdminBundle/Resources/config/services/listener.xml [hunk 5]
```diff
@@ -20,16 +20,22 @@
             <tag name="kernel.event_listener" event="sylius.shipment.post_ship" method="sendConfirmationEmail" />
         </service>
 
+<!--        <service id="sylius.event_listener.product_option_value" class="Sylius\Bundle\AdminBundle\EventListener\ProductOptionValueDeleteListener">-->
+<!--            <argument type="service" id="router" />-->
+<!--            <argument type="service" id="request_stack" />-->
+<!--            <tag name="kernel.event_listener" event="kernel.exception" method="onProductOptionValueDelete" />-->
+<!--        </service>-->
+
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/config/services/listener.xml]: I guess it should be removed 🤔

File: src/Sylius/Bundle/CoreBundle/spec/EventListener/ResourceDeleteListenerSpec.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+use PhpSpec\ObjectBehavior;
+use Sylius\Component\Core\Exception\ResourceDeleteException;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\ExceptionEvent;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+use Symfony\Component\HttpKernel\KernelInterface;
+
+final class ResourceDeleteListenerSpec extends ObjectBehavior
+{
+    function it_should_do_nothing_if_exception_is_not_foreign_key_constraint(KernelInterface $kernel): void
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/spec/EventListener/ResourceDeleteListenerSpec.php]:
```suggestion
    function it_does_nothing_if_exception_is_not_foreign_key_constraint(KernelInterface $kernel): void
```

File: src/Sylius/Bundle/AdminBundle/EventListener/ResourceDeleteExceptionListener.php [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        FlashBagProvider::getFlashBag($this->requestStack)->add('error', [
+            'message' => 'sylius.resource.delete_error',
+            'parameters' => ['%resource%' => $exception->getResourceName()],
+        ]);
+
+        $eventRequest = $event->getRequest();
+        $requestAttributes = $eventRequest->attributes;
+        $originalRoute = $requestAttributes->get('_route', '');
+
+        $referrer = $eventRequest->headers->get('referer');
```
Comment (mpysiak) [inline src/Sylius/Bundle/AdminBundle/EventListener/ResourceDeleteExceptionListener.php:49]: Is it necessary to depend on the referrer? This often raises security concerns.

**Review by mpysiak — DISMISSED** [review]

File: src/Sylius/Bundle/AdminBundle/EventListener/ResourceDeleteListener.php [hunk 8]
```diff
@@ -14,31 +14,12 @@
...
-use Symfony\Component\HttpFoundation\RedirectResponse;
+use Sylius\Component\Core\Exception\ResourceDeleteException;
 use Symfony\Component\HttpFoundation\Request;
-use Symfony\Component\HttpFoundation\RequestStack;
 use Symfony\Component\HttpKernel\Event\ExceptionEvent;
-use Symfony\Component\HttpKernel\KernelEvents;
-use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
 
-final readonly class ResourceDeleteSubscriber implements EventSubscriberInterface
+final readonly class ResourceDeleteListener
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/EventListener/ResourceDeleteListener.php:21]: Add a note about this change in the UPGRADE file

File: src/Sylius/Bundle/AdminBundle/spec/EventListener/ResourceDeleteListenerSpec.php [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    function it_does_nothing_if_request_is_not_main(KernelInterface $kernel, ForeignKeyConstraintViolationException $exception): void
+    {
+        $event = new ExceptionEvent($kernel->getWrappedObject(), new Request(), HttpKernelInterface::SUB_REQUEST, $exception->getWrappedObject());
+
+        $this->onResourceDelete($event)->shouldReturn(null);
+    }
+
+    function it_should_throw_resource_delete_exception_if_all_conditions_are_met(KernelInterface $kernel, ForeignKeyConstraintViolationException $exception): void
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/spec/EventListener/ResourceDeleteListenerSpec.php:38]:
```suggestion
    function it_throws_resource_delete_exception_if_all_conditions_are_met(KernelInterface $kernel, ForeignKeyConstraintViolationException $exception): void
```
and the same applies to other methods in this spec

File: src/Sylius/Bundle/ApiBundle/StateProcessor/Admin/ShippingMethod/RemoveProcessor.php [hunk 10]
```diff
@@ -18,6 +18,7 @@
 use ApiPlatform\State\ProcessorInterface;
 use Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException;
 use Sylius\Bundle\ApiBundle\Exception\ShippingMethodCannotBeRemoved;
+use Sylius\Component\Core\Exception\ResourceDeleteException;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/StateProcessor/Admin/ShippingMethod/RemoveProcessor.php:21]:
```suggestion
```
and the same applies to the below classes

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductVariantRepository.php [hunk 11]
```diff
@@ -64,4 +65,14 @@ public function createCatalogPromotionListQueryBuilder(
...
     }
+
+    public function countByProductOptionValueId(mixed $id): int
+    {
+        return (int) $this->createQueryBuilder('o')
+            ->select('COUNT(o.id)')
+            ->innerJoin('o.optionValues', 'optionValue', Join::WITH, 'optionValue.id = :id')
+            ->setParameter('id', $id)
+            ->getQuery()
+            ->getSingleScalarResult();
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductVariantRepository.php:76]:
```suggestion
            ->getSingleScalarResult()
        ;
```

**Review by GSadee — APPROVED** [review]

**Pull request merged**